File: crmctl.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* -- commands
 *
*/


int showHelp () {
    char help[4096] = "\ncrmctl: simple cluster management tool\n"
                      "\nCommands:\n"
                      "\n  State Overview:"
                      "\n    crmctl state"
                      "\n\n  Locate:\n"
                      "    locate <resource>\n    locate master <resource>\n"
                      "\n  Manage:\n"
                      "    clean <resource> [ <node> ]\n\n    ban <resource> <node>\n    unban <resource> <node>\n\n"
                      "    disable <resoource>\n    enable <resource> \n\n  Failover:\n    failover <resource>\n"
                      "\n  Query:"
                      "\n    nodes [ ips ]\n    resources [ brief ]\n    config\n    constraints\n    properties\n";

    printf("%s", help);
    return 0;

}

/* --- find where a resource is running--- */
int locateResource (char *resource)
{
    // replaces: pcs resource --full | grep $RESOURCE
    char cmd[1024];
    snprintf(cmd, sizeof(cmd), "crm_resource --locate --resource %s", resource);
    return system(cmd);
}

int locateResourceMaster (char *resource)
{
    // replaces: pcs status --full | grep Master | grep $RESOURCE
    char cmd[1024];
    snprintf(cmd, sizeof(cmd), "crm_resource --locate --resource %s --master", resource);
    return system(cmd);
}

/* --- ban / clear --- */
int banResourceOnNode (char *resource, char *node)
{
    // replaces: pcs resource ban $RESOURCE $NODE
    char cmd[1024];
    snprintf(cmd, sizeof(cmd), "crm_resource --resource %s --ban --node %s --quiet", resource, node);
    return system(cmd);
}

int unbanResourceOnNode (char *resource, char *node )
{
    // replaces: pcs resource cleanup $RESOURCE $NODE
    char cmd[1024];
    snprintf(cmd, sizeof(cmd), "crm_resource --resource %s --clear --node %s --quiet", resource, node);
    return system(cmd);
}

/* --- pcs cleanup replacement --- */
int cleanResourceOnNode (char *resource, char *node)
{
    // replaces: pcs resource cleanup $RESOURCE $NODE
    char cmd[1024];
    snprintf(cmd, sizeof(cmd), "crm_resource --resource %s --cleanup --node %s --force --quiet", resource, node);
    return system(cmd);
}

int cleanResource (char *resource, int force)
{
    // replaces: pcs resource cleanup $RESOURCE
    char cmd[1024];
    char force_cmd[1024];
    if (force == 1)
    {
        strcpy(force_cmd, "--force");
    } else
    {
        strcpy(force_cmd, "");
    }
    snprintf(cmd, sizeof(cmd), "crm_resource --resource %s --cleanup %s --quiet", resource, force_cmd);
    return system(cmd);
}

/* --- enable, disable resource --- */
int enableResource (char *resource, int start)
{
    // replaces: pcs resource enable  [ --start ]
    char cmd[1024];
    if (start == 1)
    {
        snprintf(cmd, sizeof(cmd), "pcs resource enable %s --start", resource);
    }
    else
    {
        snprintf(cmd, sizeof(cmd), "pcs resource enable %s", resource);
    }
    return system(cmd);
}

int disableResource ( char *resource, int force)
{
    // replaces: pcs resource disable [ --force ]
    char cmd[1024];
    if (force == 1)
    {
        snprintf(cmd, sizeof(cmd), "pcs resource disable %s", resource);
    }
    else
    {
        snprintf(cmd, sizeof(cmd), "pcs resource disable %s --force", resource);
    }
    return system(cmd);
}

/* failover resource */
int failoverResource ( char *resource )
{
    // replaces: /usr/local/eptt/bin/failover.sh RESOURCE INT INT INT INT
    char cmd[1024];
    snprintf(cmd, sizeof(cmd), "/usr/local/eptt/bin/monitored-failover.sh -t %s -m", resource);
    return system(cmd);
}

int moveResource (char *resource )
{
    printf("Can't move resource - not implemented!");
    return 127;
}

/* get overviews, statii, etc */
int getNodes ()
{
    // replaces: pcs status --full | grep Online | cut -d '[' -f 2 | cut -d ']' -f 1 | xargs
    return system("crm_node --list | xargs");
}

int getNodeHosts ()
{
    // replaces: ?
    char cmd[2048] = "cat /etc/hosts | grep -v localhost | grep -v '::' | grep -v 172 | cut -d ' ' -f 1,2 | grep -E 'ptt.*|priv.*|web'";
    return system(cmd);
}

int getResources( int brief)
{
    if (brief == 1)
    {

        // replaces: pcs resource show $RESOURCES
        return system("crm_resource --list-raw | cut -d ':' -f 1 | sort -u");
    }
    else
    {
        // replaces: pcs resource show --full
        return system("crm_resource --list-raw");
    }
}

int getState ()
{
    //replaces: pcs status --full
    return system("crm_mon --one-shot --show-detail");
}

int getConfig ()
{
    return system("pcs config --full");
}

int getProperties ()
{
    return system("pcs property show --full");
}

int getAllConstraints ()
{
    //replaces: pcs constraint show --full
    return system("crm_mon --neg-locations --show-detail --one-shot");
}

int getBlame () {
    /* TODO: remove this */
    printf("\ncrmctl v1.0\nIf this is broken, blame James #4\n");
    return 1;
}
int main(int argc, char *argv[])
{
    /* Commands are evaluated and executed in the following order:
     * - [ missing argv[1] ]
     * - state
     * - nodes
     * - resources
     * - constraints
     * - config
     * - blame
     * - properties
     * - locate
     * - ban
     * - unban
     * - clean
     * - disable
     * - enable
     * - failover
     * - [ invalid argv[1], have[ argv[2,3,4] ]]
     *
     * Exit code is 0 for success, 1 for failure.
     *
     */

    // Will be our return code.
    unsigned int r;

    if (argc == 1)
    {
        showHelp();
        printf("\nFATAL: At least one argument is required!\n");
        return 1;
    }

    /* Process args */

    // STATE
    if (strcmp(argv[1], "state") == 0)
    {
        r = getState();
    }
        // NODES
    else if (strcmp(argv[1], "nodes") == 0)
    {
        if (argc == 2)
        {
            r = getNodes();
        } else if (argc == 3)
        {
            r = getNodeHosts();
        } else
        {
            printf("Fatal: invalid arg passed to: crmctl nodes [ ips ]\n");
            return 1;
        }
    }
        // RESOURCES
    else if (strcmp(argv[1], "resources") == 0)
    {
        if (argc == 3)
        {
            if (strcmp(argv[2], "brief") == 0)
            {
                r = getResources(1);
            }
        }
        else
        {
                r = getResources(0);
        }
    }
        // CONSTRAINTS
    else if (strcmp(argv[1], "constraints") == 0)
    {
        r = getAllConstraints();
    }
        // CONFIG
    else if (strcmp(argv[1], "config") == 0)
    {
        r = getConfig();
    }
        // BLAME
    else if (strcmp(argv[1], "blame") == 0)
    {
        r = getBlame();
    }
        // PROPERTIES
    else if (strcmp(argv[1], "properties") == 0)
    {
        r = getProperties();
    }
        // LOCATE | FIND
    else if ((strcmp(argv[1], "locate") == 0) || (strcmp(argv[1], "find") == 0))
    {
        if (strcmp(argv[2], "master") == 0)
        {
            printf("Locate Master");
            r = locateResourceMaster(argv[2]);
        } else
        {
            printf("Locate Resource");
            r = locateResource(argv[2]);
        }
    }
        // BAN
    else if (strcmp(argv[1], "ban") == 0 && strcmp(argv[3], "") != 0)
    {
        printf("BAN\n");
        r = banResourceOnNode(argv[2], argv[3]);
    }
        // UNBAN/CLEAR
    else if (strcmp(argv[1], "unban") == 0 && strcmp(argv[3], "") != 0)
    {
        printf("Unban\n");
        r = unbanResourceOnNode(argv[2], argv[3]);
    }
        // CLEAN
    else if (strcmp(argv[1], "clean") == 0)
    {
        if (argc == 4)
        {
            printf("Cleanup %s on %s\n", argv[2], argv[3]);
            r = cleanResourceOnNode(argv[2], argv[3]);
        } else if (argc == 3)
        {
            printf("Cleanup %s\n", argv[2]);
            r = cleanResource(argv[2], 1);
        } else
        {
            printf("crmctl cleanup <resource> [ <NODE> ]\nFatal: invalid usage\n");
            r = 1;
        }
    }
        // DISABLE
    else if (strcmp(argv[1], "disable") == 0 && strcmp(argv[2], "") != 0)
    {
        printf("Disable %s\n", argv[2]);
        r = disableResource(argv[2], 1);
    }
        // ENABLE
    else if (strcmp(argv[1], "enable") == 0 && strcmp(argv[2], "") != 0)
    {
        printf("Enable %s\n", argv[2]);
        r = enableResource(argv[2], 1);
    }
        // FAILOVER
    else if (strcmp(argv[1], "failover") == 0 && strcmp(argv[2], "") != 0)
    {
        printf("Failover resource: %s\n", argv[2]);
        r = failoverResource(argv[2]);
    }
        // HELP
    else if (strcmp(argv[1], "help") == 0)
    {
        r = showHelp();
    }
        // BAD SECONDARY INPUT
    else
    {
        /* if we dont have an action (first arg), then it is caught at top */
        if (argc == 2)
        {
            showHelp();
            printf("\nFATAL: invalid action %s\n", argv[1]);
        } else if (argc == 3)
        {
            showHelp();
            printf("FATAL: Bad command or substitution: %s -> %s\n", argv[1], argv[2]);
        } else
        {
            printf("FATAL: Bad or unknown command provided.\n");
        }
        r = 1;
    }
    return r;
}

File: crmctl.py
#!/usr/bin/env python3.6

import argparse
import sys
import os

# Assuming we installed via RPM or by the ./install.sh script, should already be there:

sys.path.insert(0, "/etc/crmctl")
sys.path.append("/etc/crmctl/lib")

# Import the crmctl libs
import lib


help = """
Usage: crmctl <command> [ <subcommand1>, <subcommand2>, ... ]

Commands:

  State Overview:                  crmctl state

  Locate:                          crmctl locate [ master ] <resource>

  Manage:
    Cleanup:                       crmctl clean <resource> [ <node>, all ]
    Ban:                           crmctl ban <resource> [ <node> ]
    Unban:                         crmctl unban <resource> [ <node> ]
    Disable:                       crmctl disable <resource> [ <node> ] 
    Enable:                        crmctl enable <resource> [ <node> ]
    Remote Execute:                crmctl node-exec <node> <command>

  Failover:
    Failover:                      crmctl failover <resource>
    Staggered Restart:             crmctl stagger-restart <resource>
    Transit Masters to Node:       crmctl transit-resources <node>

  Query:
    Cluster Nodes:                 crmctl nodes [ ips ]
    Cluster Resources:             crmctl resources [ <resource> ]
    Cluster Configuration:         crmctl config
    Resource Constraints:          crmctl constraints
    Cluster Properties:            crmctl properties
    EPTT Service Console:          crmctl console <service,mgwd,sipd> <command>
                                   crmctl remote-console <node> <service> <command>

  Server Logs:                     crmctl logs <log_type> [ [ <resource_selector> ], [ <num_lines> ] ]

  Cluster Information Base (CIB):  crmctl tree [ <section> ]
"""


class CommandProcessor:

    def __init__(self):
        self.parser = argparse.ArgumentParser()
        self.configure_args()
        self.process()

    def configure_args(self):

        self.parser.add_argument(
            "command",
            action="store",
            nargs="*"
        )

    def process(self):
        """ Process a crmctl request."""

        args = self.parser.parse_args()

        def pos(x):
            """
            command structure:
                    popped off -> target
                    VVV
            crmctl command sub third fourth [ *args ]
                            |    |     |
                            0    1     2

            """
            if len(args.command) > x:
                return args.command[x].strip()
            else:
                return ""

        # We got no command; ie `crmctl `
        if len(args.command) == 0:
            failed = "Fatal: No command provided to crmctl. For a list of commands, 'crmctl help'"
            action = ""
        else:
            # We got a command: take this, and turn it into the action to perform, removing it from the 0 position
            # and shifting all commands one position to the left
            action = args.command.pop(0)

        # Subcommand:
        sub = pos(0)

        # Third position, used for optional tertiary commands
        third = pos(1)

        # Fourth position, used for optional quaternary commands
        fourth = pos(2)

        # ... everything _after_ this that gets passed into crmctl is iterated over with pos(x)

        # How many default log lines?
        log_lines = None

        # Load /etc/crmctl/crmctl.conf
        conf_loader = lib.ConfigLoader()
        conf = conf_loader.load_conf()

        if conf:
            log_lines = conf.get("default_maximum_log_lines")
            version = conf.get("version")
            if not version:
                version = 1


        cib = lib.CIBTree()
        crm = lib.CRMResource()
        output_printers = lib.FormattedOutputs()
        console = lib.Console()

        failed = False
        cmd_return = None

        _invalid_args = "Invalid number of args passed; expected either <resource> , or <resource> <node>"

        # - crmctl version
        if action in ["version"]:
            cmd_return = "crmctl: v0.{}".format(version)
            print(cmd_return)

        # - crmctl help
        if action in ["h", "help"]:
            cmd_return = help
            print(cmd_return)

        #  - crmctl clean [ <resource [ <node1>, <node2> ]]
        if action in ["clean", "cn"]:
            if sub == "all":
                cmd_return = crm.cleanup_all()
            elif sub and not third:
                cmd_return = crm.cleanup_resource(sub)
            elif sub:
                for i in range(1, len(args.command)):
                    cmd_return = crm.cleanup_resource_on_node(sub, args.command[i])
            else:
                failed = _invalid_args

        # - crmctl ban <resource> [ <node> ]
        elif action == "ban":
            if sub and third:
                cmd_return = crm.ban_resource_on_node(sub, third)
            elif sub:
                cmd_return = crm.ban_resource(sub)
            else:
                failed = _invalid_args

        # - crmctl unban <resource> [ <node> ]
        elif action == "unban":
            if sub and third:
                cmd_return = crm.unban_resource_on_node(sub, third)
            elif sub:
                cmd_return = crm.unban_resource(sub)
            else:
                failed = _invalid_args

        # - crmctl state
        elif action in ["status", "state", "st"]:
            cmd_return = crm.get_state()

        # - crmctl nodes
        elif action in ["nd", "nodes"]:
            if sub and sub in ["ip", "i", "ips"]:
                cmd_return = crm.list_nodes_and_ips()
            else:
                cmd_return = crm.list_nodes()

        # - crmctl config
        elif action in ["cf", "cnf", "config", "cfg"]:
            cib.get_tree("configuration")
            cmd_return = cib.get_service_configuration()

        # - crmctl properties
        elif action in ["properties", "prop", "pt", "pr"]:
            cmd_return = cib.get_cluster_properties()

        # - crmctl constraints
        elif action in ["constraints", "const", "cnst", "constraint"]:
            cmd_return = output_printers.print_brief_constraints_by_node(sub)

            if sub:
                for frame in cib.get_constraint_attrib_iter():
                    try:
                        for key, value in frame.items():
                            print("{:20s}: {:20s}".format(str(key), str(value)))
                    except:
                        continue

        # - crmctl stagger-restart
        elif action in ["stagger-restart", "stagger", "sf"]:
            if action in ["stagger"] and not sub == "restart":
                cmd_return = "Stagger? restart, reload"
                print(cmd_return)
            else:
                if sub:
                    remaining_args = "/usr/local/eptt/bin/stagger-restart-resource.sh {}".format(sub)
                    cmd_return = lib.LocalExec(remaining_args)
                else:
                    failed = "Invalid command: need target to stagger-restart."

        # - crmctl resources
        elif action in ["r", "rs", "rsc", "resources"]:
            if sub:
                cmd_return = output_printers.print_basic_resource_overview(sub)
            else:
                cmd_return = crm.list_resources()

        # - crmctl tree
        elif action in ["tree", "xt", "xtree"]:
            cmd_return = output_printers.print_cib_tree()

        # - crmctl logs [ <logfilename> ]
        elif action in ["logs", "log"]:
            if sub in ["messages", "docker"]:
                if third and not fourth:
                    fourth = third
                    third = None
                elif third and log_lines:
                    fourth = log_lines

            reader = lib.tools.LogReader(sub, third, num_lines=fourth)
            good_return, logs = reader.readlines()

            if not good_return:
                print(logs)
            else:
                for line in logs:
                    print(line)
            cmd_return = logs

        # - crmctl console
        # - crmctl remote-console
        elif action in ["console", "con", "cns", "remote-console", "rcon", "rcns"]:
            if not third:
                failed = "Fatal: crmctl console <service> [ sum, lic, <console command> ]"
            else:
                if action in ["remote-console", "rcon", "rcns"]:
                    node = sub
                    if not node in crm.list_nodes():
                        failed = "Fatal: crmctl remote-console <node> <service> <command> - did you forget to specify the node?"
                        print(failed)
                        start = 0
                    else:
                        sub = third
                        start = 2
                else:
                    node = None
                    start = 1

                if not failed:
                    remaining_args = []
                    for i in range(start, len(args.command)):
                        remaining_args.append(pos(i))
                    cmd_return = console.console(sub, " ".join(remaining_args), node=node)
                    if not cmd_return.strip().strip("\n"):
                        failed = "No response from console '{}' - is this a valid console target?".format(sub)

        # - crmctl locate <resource>
        # - crmctl locate master <resource>
        elif action in ["loc", "find", "locate"]:
            if sub and sub == "master" and third:
                cmd_return = crm.locate_resource_master(sub)
            elif sub:
                cmd_return = crm.locate_resource(sub)
            else:
                failed = _invalid_args

        # - crmctl failover <resource>
        elif action == "failover":
            if sub:
                if os.path.exists("/usr/local/eptt/bin/monitored-failover.sh"):
                    if sub:
                        cmd_return = lib.LocalExec("/usr/local/eptt/bin/monitored-failover.sh -t {} -m".format(sub))
                    else:
                        print("Do not have monitored-failover.sh!")
                        failed = _invalid_args
                else:
                    failed = help + "Path to monitored-failover does not exist!"
            else:
                failed = "Missing target for monitored failover - cannot continue."

        # - crmctl shift <node>
        elif action in ["transit-resources", "transit-resource", "shift", "fail-to", "failall", "shf", "transit"]:
            if not sub:
                failed = "Fatal: crmctl {} <node>; missing node to fail all master/slave sets too.".format(action)
            elif sub:
                sure = "".join(["WARN: You are about to fail every cluster resource over to {}\n".format(sub),
                                "This may result in multiple failover actions.",
                                "All services will be shifted to node {}. Continue? [N]".format(
                                    sub
                                )])
                yn = input(sure)
                if yn in ["y", "Y", "yes", "Yes"]:
                    cmd_return = lib.LocalExec("/usr/local/eptt/bin/monitored-failover.sh -N {}".format(sub))
            else:
                failed = "Fatal: crmctl {} : cannot fail all resource to node specified - invalid or msisings args."

        # - crmctl node-run <node> <command>
        elif action in ["node-exec", "node-run", "nde", "ndx"]:
            if not sub in crm.list_nodes():
                "Fatal: crmctl node-exec <node> <command> - invalid or unspecified node: {}".format(sub)
            elif not third:
                failed = "Fatal: missing command to execute on node!"
            else:
                to_run = " ".join([args.command[c] for c in range(1, len(args.command))])
                cmd_return = lib.SSHExec(sub, to_run)

        # - Invalid number of args provided, or bad arg combination.
        if failed:
            print(failed)
            print("Fatal: invalid command configuration : {}".format(
                ", ".join([action, ", ".join([c for c in args.command])])
                )
            )

        if not cmd_return:
            print(help)
            passed_in = " ".join(args.command)
            if not action:
                print("Fatal: crmctl requires at least one argument.")
            else:
                print("Fatal: invalid configuration or unknown command given to crmctl: '{}'".format(
                    " ".join([action, passed_in]))
                )


if __name__ == "__main__":
    CommandProcessor()
